python-lib/data_loader.py

```diff
@@ -39,37 +39,50 @@ def _validate_df(self, df: pd.DataFrame):
             logging.info(
                 f"Loading dataframe of {len(df.index)} rows and "
-                + f"{len(self.feature_columns)} column(s) into vector format..."
+                + f"{len(self.feature_columns)} column(s) into array format..."
             )
-        vector_ids = df[self.unique_id_column].values
-        vectors = np.empty(shape=(len(df.index), self.MAX_VECTOR_LENGTH))  # pre-allocate empty array of fixed size
+        array_ids = df[self.unique_id_column].values
+        array_length = self._count_array_length(df)
+        arrays = self._load_arrays_from_df(df, array_length)
+        if verbose:
+            logging.info(
+                f"Loading dataframe into array format: dimensions {arrays.shape} "
+                + f"loaded in {perf_counter() - start:.2f} seconds.",
+            )
+        return (array_ids, arrays)
+
+    def _load_arrays_from_df(self, df: pd.DataFrame, array_length: int) -> np.array:
+        """Concatenate numeric/array columns of DataFrame into a single numpy.array"""
+        arrays = np.empty(shape=(len(df.index), array_length))  # pre-allocate empty array of fixed size
         i = 0
         for column in self.feature_columns:
-            column_is_vector = df[column].dtype == "object" and df[column].str.startswith("[").all()
-            if column_is_vector:
+            column_is_array = df[column].dtype == "object" and df[column].str.startswith("[").all()
+            if column_is_array:
                 try:
-                    column_array = np.stack(df[column].apply(self.load_vector_from_string), axis=0)
-                    vectors[:, i : (i + column_array.shape[1])] = column_array  # noqa
+                    column_array = np.stack(df[column].apply(self.load_array_from_string), axis=0)
+                    arrays[:, i : (i + column_array.shape[1])] = column_array  # noqa
                     i += column_array.shape[1]
                 except ValueError as e:
-                    raise ValueError(f"Invalid vector data in column '{column}': {e}")
+                    raise ValueError(f"Invalid array data in column '{column}': {e}")
             else:
                 try:
-                    vectors[:, i] = df[column].values.astype(np.float32)
+                    arrays[:, i] = df[column].values.astype(np.float32)
                     i += 1
                 except ValueError as e:
                     raise ValueError(f"Invalid numeric data in column '{column}': {e}")
-        vectors = np.ascontiguousarray(vectors[:, :i], dtype=np.float32)
-        if verbose:
-            logging.info(
-                f"Loading dataframe into vector format: array of dimensions {vectors.shape} "
-                + f"loaded in {perf_counter() - start:.2f} seconds.",
-            )
-        return (vector_ids, vectors)
+        arrays = np.ascontiguousarray(arrays[:, :i], dtype=np.float32)
+        return arrays
+
+    def _count_array_length(self, df: pd.DataFrame) -> int:
```

Comment:
For the function _count_array_length, what is the first array here, is it not supposed to be the only one we get ?

Reply:
Ah ok this it the first row of dataset